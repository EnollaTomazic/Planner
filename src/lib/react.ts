import * as React from "react";

export const hasTextContent = (node: React.ReactNode): boolean => {
  if (node === null || node === undefined) return false;
  if (typeof node === "boolean") return false;
  if (typeof node === "string") return node.trim().length > 0;
  if (typeof node === "number") return true;
  if (Array.isArray(node)) return node.some((item) => hasTextContent(item));
  if (React.isValidElement<{ children?: React.ReactNode }>(node)) {
    return hasTextContent(node.props.children ?? null);
  }
  return false;
};

export function observeMediaQuery(
  query: string,
  onChange: (matches: boolean, mediaQuery: MediaQueryList) => void,
): (() => void) | undefined {
  if (typeof window === "undefined" || typeof window.matchMedia !== "function") {
    return undefined;
  }

  let mediaQuery: MediaQueryList;

  try {
    mediaQuery = window.matchMedia(query);
  } catch {
    return undefined;
  }

  const handleChange = (event: MediaQueryListEvent): void => {
    onChange(event.matches, mediaQuery);
  };

  const handleLegacyChange = (): void => {
    onChange(mediaQuery.matches, mediaQuery);
  };

  onChange(mediaQuery.matches, mediaQuery);

  if (typeof mediaQuery.addEventListener === "function") {
    mediaQuery.addEventListener("change", handleChange);
    return () => {
      mediaQuery.removeEventListener("change", handleChange);
    };
  }

  if (typeof mediaQuery.addListener === "function") {
    mediaQuery.addListener(handleLegacyChange);
    return () => {
      mediaQuery.removeListener(handleLegacyChange);
    };
  }

  return undefined;
}

export function useMatchMedia(
  query: string,
  getServerSnapshot?: () => boolean,
): boolean {
  const getMatches = React.useCallback(() => {
    if (typeof window === "undefined") {
      return typeof getServerSnapshot === "function" ? getServerSnapshot() : false;
    }

    if (typeof window.matchMedia !== "function") {
      return false;
    }

    try {
      return window.matchMedia(query).matches;
    } catch {
      return false;
    }
  }, [query, getServerSnapshot]);

  const [matches, setMatches] = React.useState<boolean>(getMatches);

  React.useEffect(() => {
    return observeMediaQuery(query, (nextMatches) => {
      setMatches(nextMatches);
    });
  }, [query]);

  return matches;
}
