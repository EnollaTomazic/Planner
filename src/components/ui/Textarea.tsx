// src/components/ui/Textarea.tsx
"use client";

import * as React from "react";
import { cn } from "@/lib/utils";

export interface TextareaProps
  extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  tone?: "default" | "pill";
  helperText?: string;
  errorText?: string;
  success?: boolean;
  leadingIcon?: React.ReactNode;
  trailingAction?: React.ReactNode;
}

const Textarea = React.forwardRef<HTMLTextAreaElement, TextareaProps>(function Textarea(
  {
    className,
    helperText,
    errorText,
    success,
    leadingIcon,
    trailingAction,
    disabled,
    id,
    tone = "default",
    ...props
  },
  ref
) {
  function slug(s?: string) {
    return (s ?? "")
      .toLowerCase()
      .trim()
      .replace(/\s+/g, "-")
      .replace(/[^a-z0-9\-]/g, "")
      .slice(0, 64);
  }

  const auto = React.useId();
  const fromAria = slug(props["aria-label"] as string | undefined);
  const finalId = id || auto;
  const finalName = props.name || fromAria || finalId;
  const describedBy = errorText ? `${finalId}-error` : helperText ? `${finalId}-helper` : undefined;
  return (
    <div className="space-y-1">
      <div
        className={cn(
          "group relative inline-flex w-full items-start rounded-[16px] border border-[hsl(var(--border)/0.25)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out before:content-[''] before:scanlines before:opacity-5 after:content-[''] after:noise",
          "hover:border-[hsl(var(--border)/0.35)] hover:shadow-[0_0_0_1px_hsl(var(--border)/0.2)]",
          "focus-within:ring-2 focus-within:ring-[hsl(var(--accent)/0.28)] focus-within:ring-offset-2 focus-within:ring-offset-[hsl(var(--background))] focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] jitter",
          tone === "pill" && "rounded-full min-h-[120px]",
          errorText && "border-[hsl(var(--destructive)/0.55)] focus-within:ring-[hsl(var(--destructive)/0.35)]",
          success && "border-[hsl(var(--success)/0.45)] focus-within:ring-[hsl(var(--success)/0.28)]",
          (disabled || props.readOnly) && "opacity-60 cursor-not-allowed focus-within:ring-0 focus-within:shadow-none",
          className
        )}
      >
        {leadingIcon && (
          <span className="pl-3.5 pt-3 text-[hsl(var(--muted-foreground))] group-focus-within:text-[hsl(var(--accent))] flex items-start">
            {leadingIcon}
          </span>
        )}
        <textarea
          ref={ref}
          id={finalId}
          name={finalName}
          disabled={disabled}
          aria-invalid={errorText ? "true" : props["aria-invalid"]}
          aria-describedby={describedBy}
          className={cn(
            "flex-1 min-h-[96px] h-24 px-3.5 py-3 text-sm bg-transparent placeholder:text-[hsl(var(--muted-foreground))/0.8] text-[hsl(var(--foreground))] caret-[hsl(var(--accent))] outline-none resize-y disabled:cursor-not-allowed"
          )}
          {...props}
        />
        {trailingAction && (
          <span className="pr-3.5 pt-3 text-[hsl(var(--muted-foreground))] group-focus-within:text-[hsl(var(--accent))] flex items-start">
            {trailingAction}
          </span>
        )}
      </div>
      {(helperText || errorText) && (
        <p
          id={errorText ? `${finalId}-error` : helperText ? `${finalId}-helper` : undefined}
          className={cn(
            "text-xs mt-1 line-clamp-2",
            errorText ? "text-[hsl(var(--destructive))]" : "text-[hsl(var(--muted-foreground))]"
          )}
        >
          {errorText || helperText}
        </p>
      )}
    </div>
  );
});

export default Textarea;
