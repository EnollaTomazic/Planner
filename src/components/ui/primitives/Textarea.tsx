"use client";

import * as React from "react";
import { cn, slugify } from "@/lib/utils";
import FieldShell from "./FieldShell";

export type TextareaProps = React.TextareaHTMLAttributes<HTMLTextAreaElement> & {
  /** Rounded look without needing a global .planner-textarea class. */
  tone?: "default" | "pill";
  /** Optional className for the outer wrapper */
  className?: string;
  /** Optional className for the inner <textarea> element */
  textareaClassName?: string;
};

  const INNER =
    "block w-full max-w-full min-h-7 px-4 py-3 text-base bg-transparent " +
  "text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))] " +
  "focus:outline-none focus-visible:outline-none resize-y disabled:opacity-50 disabled:cursor-not-allowed";

export default React.forwardRef<HTMLTextAreaElement, TextareaProps>(function Textarea(
  {
    className,
    textareaClassName,
    id,
    name,
    "aria-label": ariaLabel,
    tone = "default",
    ...props
  },
  ref
) {
  const auto = React.useId();
  const fromAria = slugify(ariaLabel as string | undefined);
  // Use React-generated id by default so multiple fields sharing an aria-label
  // do not end up with duplicate ids.
  const finalId = id || auto;
  const finalName = name || fromAria || finalId;

  const error =
    props["aria-invalid"] === true || props["aria-invalid"] === "true";

  return (
    <FieldShell
      tone={tone}
      error={error}
      disabled={props.disabled}
      className={className}
    >
      <textarea
        ref={ref}
        id={finalId}
        name={finalName}
          className={cn(INNER, tone === "pill" && "rounded-full min-h-6", textareaClassName)}
        {...props}
      />
    </FieldShell>
  );
});
