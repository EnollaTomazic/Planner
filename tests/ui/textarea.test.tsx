import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { describe, it, expect, afterEach } from 'vitest';
import Textarea from '../../src/components/ui/primitives/Textarea';

afterEach(cleanup);

describe('Textarea', () => {
  it('renders default state', () => {
    const { container } = render(<Textarea aria-label="test" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders focus state', () => {
    const { container, getByRole } = render(<Textarea aria-label="test" />);
    fireEvent.focus(getByRole('textbox'));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders disabled state', () => {
    const { container } = render(<Textarea aria-label="test" disabled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders pill tone', () => {
    const { container } = render(<Textarea aria-label="test" tone="pill" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders error state', () => {
    const { container } = render(
      <Textarea aria-label="test" aria-invalid="true" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('handles aria-invalid="false" as non-error', () => {
    const { container } = render(
      <Textarea aria-label="test" aria-invalid="false" />
    );
    expect(container.firstChild).not.toHaveClass(
      'border-[hsl(var(--destructive)/0.6)]'
    );
  });

  it('has no outline when focused', () => {
    const { getByRole } = render(<Textarea aria-label="outline" />);
    const ta = getByRole('textbox');
    fireEvent.focus(ta);
    const style = getComputedStyle(ta);
    expect(style.outlineStyle === 'none' || style.outlineStyle === '').toBe(true);
    expect(style.outlineWidth === '0px' || style.outlineWidth === '').toBe(true);
  });
});
