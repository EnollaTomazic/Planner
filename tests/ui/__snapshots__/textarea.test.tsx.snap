// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Textarea > renders default state 1`] = `
<div
  class="relative inline-flex w-full items-center rounded-xl border border-[hsl(var(--border)/0.28)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-[--theme-ring] focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,rgba(195,255,255,0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')]"
>
  <textarea
    class="block w-full max-w-full min-h-40 px-4 py-3 text-base bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))] focus:outline-none focus-visible:outline-none resize-y disabled:opacity-50 disabled:cursor-not-allowed"
    id=":r0:"
    name="test"
  />
</div>
`;

exports[`Textarea > renders disabled state 1`] = `
<div
  class="relative inline-flex w-full items-center rounded-xl border border-[hsl(var(--border)/0.28)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-[--theme-ring] focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,rgba(195,255,255,0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')] opacity-60 pointer-events-none"
>
  <textarea
    class="block w-full max-w-full min-h-40 px-4 py-3 text-base bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))] focus:outline-none focus-visible:outline-none resize-y disabled:opacity-50 disabled:cursor-not-allowed"
    disabled=""
    id=":r2:"
    name="test"
  />
</div>
`;

exports[`Textarea > renders error state 1`] = `
<div
  class="relative inline-flex w-full items-center rounded-xl border border-[hsl(var(--border)/0.28)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-[--theme-ring] focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,rgba(195,255,255,0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')] border-[hsl(var(--destructive)/0.6)] focus-within:ring-[hsl(var(--destructive)/0.35)]"
>
  <textarea
    aria-invalid="true"
    class="block w-full max-w-full min-h-40 px-4 py-3 text-base bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))] focus:outline-none focus-visible:outline-none resize-y disabled:opacity-50 disabled:cursor-not-allowed"
    id=":r4:"
    name="test"
  />
</div>
`;

exports[`Textarea > renders focus state 1`] = `
<div
  class="relative inline-flex w-full items-center rounded-xl border border-[hsl(var(--border)/0.28)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-[--theme-ring] focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,rgba(195,255,255,0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')]"
>
  <textarea
    class="block w-full max-w-full min-h-40 px-4 py-3 text-base bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))] focus:outline-none focus-visible:outline-none resize-y disabled:opacity-50 disabled:cursor-not-allowed"
    id=":r1:"
    name="test"
  />
</div>
`;

exports[`Textarea > renders pill tone 1`] = `
<div
  class="relative inline-flex w-full items-center rounded-xl border border-[hsl(var(--border)/0.28)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-[--theme-ring] focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,rgba(195,255,255,0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')] rounded-full"
>
  <textarea
    class="block w-full max-w-full min-h-40 px-4 py-3 text-base bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))] focus:outline-none focus-visible:outline-none resize-y disabled:opacity-50 disabled:cursor-not-allowed rounded-full min-h-32"
    id=":r3:"
    name="test"
  />
</div>
`;
